### PR TITLE
Self remove account support

### PR DESCRIPTION
**Pull request by sonalkr132**
closes: #1295 
ezgif com-video-to-gif

### CODE REVIEW

**Comment by sonalkr132** [comment]
Show page after owner has deleted account:
screenshot from 2016-05-31 00-56-05

**Comment by dwradcliffe** [comment]
I think I'd rather use a new page instead of a modal dialog for this. We don't use modals anywhere else in the app so I'd like to be consistent.

Also can you list the gems on that page (as described in the original issue)?

**Comment by sonalkr132** [comment]
updated :+1:

**Comment by dwradcliffe** [comment]
@qrush for review too

File: app/controllers/api/v1/owners_controller.rb [hunk 1]
```diff
@@ -25,7 +25,8 @@ def create
   def destroy
     owner = @rubygem.owners.find_by_name(params[:email])
     if owner
-      if @rubygem.ownerships.find_by_user_id(owner.id).try(:destroy)
+      ownership = @rubygem.ownerships.find_by_user_id(owner.id)
```
Comment (farukaydin) [inline app/controllers/api/v1/owners_controller.rb]: I think find_by(user_id: owner.id) would be better

File: app/models/ownership.rb [hunk 2]
```diff
@@ -3,17 +3,4 @@ class Ownership < ActiveRecord::Base
   belongs_to :user
 
   validates :user_id, uniqueness: { scope: :rubygem_id }
-
-  before_destroy :keep_last_owner
```
Comment (qrush) [inline app/models/ownership.rb]: Why was this removed?

File: app/views/profiles/delete.html.erb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<% @title = t('.title') %>
+
+<% unless @only_owner_gems.empty? %>
```
Comment (qrush) [inline app/views/profiles/delete.html.erb]: Let's try to express this positively - it's a double negative and difficult to understand.

File: app/views/profiles/_rubygem.html.erb [hunk 4]
```diff
@@ -15,4 +15,10 @@
       </div>
     <% end %>
   </a>
+  <% if defined? owners %>
```
Comment (qrush) [inline app/views/profiles/_rubygem.html.erb]: Seeing `defined?` in a view is weird. Let's use `local_assigns` instead.

File: app/views/profiles/delete.html.erb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  <div class="t-body">
+    <p class="page__subheading"><%= t '.list_only_owner' %>:</p>
+  </div>
+  <div id="profile">
+    <div class="profile-list">
+      <ul>
+        <%= render :partial => 'rubygem', collection: @only_owner_gems, locals: {graph: false, owners: true}  %>
```
Comment (qrush) [inline app/views/profiles/delete.html.erb]: Mix of 1.8 style hash keys + json style - let's prefer the json style. Also, I think it can just be `<%= render 'rubygem', collection: ... %>`

File: config/routes.rb [hunk 6]
```diff
@@ -123,7 +123,12 @@
     resource :search,    only: :show
     resource :dashboard, only: :show, constraints: { format: /html|atom/ }
     resources :profiles, only: :show
-    resource :profile, only: [:edit, :update]
+    resource :profile, only: [:edit, :update] do
+      member do
+        get :delete
+        delete :destroy, as: :destroy
```
Comment (qrush) [inline config/routes.rb]: Isn't this already on the routes as `profiles#destroy` ?

File: app/models/user.rb [hunk 7]
```diff
@@ -5,14 +5,14 @@ class User < ActiveRecord::Base
 
   PERMITTED_ATTRS = [:bio, :email, :handle, :hide_email, :location, :password, :website].freeze
 
-  has_many :rubygems, through: :ownerships
+  has_many :rubygems, through: :ownerships, dependent: :destroy
```
Comment (qrush) [inline app/models/user.rb]: We definitely don't want to destroy `Rubygem` records on destroy.

File: app/models/user.rb [hunk 8]
```diff
@@ -5,14 +5,14 @@ class User < ActiveRecord::Base
...
 
-  has_many :subscribed_gems, -> { order("name ASC") }, through: :subscriptions, source: :rubygem
+  has_many :subscribed_gems, -> { order("name ASC") }, through: :subscriptions, source: :rubygem, dependent: :destroy
```
Comment (qrush) [inline app/models/user.rb]: We want to destroy subscriptions, not the gems.

**Comment by qrush** [comment]
Nice start! A few things:

1. We don't need a link to contact support about taking over a fully yanked gem. It's just possible to push a new version onto the namespace. We should perhaps explain how to do that, or link to a guide on guides.rubygems.org.
2. Let's have one final confirm via JS before going through with the final button.
3. Instead of destroying gems, the existing gems on an account should go through the `Deletion` process that yanked gems go through. Check out [Api::V1::DeletionsController](https://github.com/rubygems/rubygems.org/blob/master/app/controllers/api/v1/deletions_controller.rb) for that flow.

I know this is a lot - this isn't a feature we should take lightly. Thanks for your patience.

File: app/models/ownership.rb (hunk 2 above)
Comment (sonalkr132) [inline app/models/ownership.rb]: I am deleting `ownership` records associated with user as well. Previously it was not possible for a gem to have no owner, now with user delete possible it will be (`orphaned` gems).

I am guessing this check was in place so that users can't remove all the gem owners with `gem owner` command. I have ensure that they still won't be able to do it [here](https://github.com/rubygems/rubygems.org/pull/1301/files#diff-76999f2b6a3a61482baf0e780a052acaR29).

Comment (qrush) [inline app/models/ownership.rb]: Yep, IIRC It's in place so someone can't remove themselves and orphan the gem - not so users can't remove others. That's definitely OK to do.

We definitely want to use the `Deletion` model to remove these ownerships so we have a record of them when the user being deleted is the only owner.

There's a lot of cases here to cover 😰 

Re: that check in the `OwnersController`, that kind of check should definitely happen in the model instead.

File: app/views/profiles/delete.html.erb (hunk 3 above)
Comment (bf4) [inline app/views/profiles/delete.html.erb]: If any?

File: app/views/profiles/delete.html.erb (hunk 5 above)
Comment (sonalkr132) [inline app/views/profiles/delete.html.erb]: > I think it can just be <%= render 'rubygem', collection: ... %>

Default usage does not support `collection` and `locals`. http://api.rubyonrails.org/classes/ActionView/PartialRenderer.html#class-ActionView::PartialRenderer-label-Rendering+the+default+case

File: app/models/user.rb (hunk 7 above)
Comment (sonalkr132) [inline app/models/user.rb]: It does not delete `rubygem`, it deletes the `ownerships`. Not saying this is a credible source, however: http://makandracards.com/makandra/32175-don-t-forget-automatically-remove-join-records-on-has_many-through-associations

I have added it to `ownership` line cause it does the same thing.. I guess it is more readable.

File: config/routes.rb (hunk 6 above)
Comment (sonalkr132) [inline config/routes.rb]: If we change
```Ruby
resources :profiles, only: :show
```
to
```Ruby
resources :profiles, only: [:show, :destroy]
```
We have a path `DELETE	/profiles/:id(.:format)` which not really the format rest of pages follow(for ex we have `/profile`, `/profile/edit`)

Comment (dwradcliffe) [inline config/routes.rb]: If you change _this_ `resource :profile, only: [:edit, :update]` to `resource :profile, only: [:edit, :update, :destroy]` it will work the way we want.

Comment (sonalkr132) [inline config/routes.rb]: Not really. Of course thats the first thing I tried. Following is what it generates:
```
edit_profile_path	GET	/profile/edit(.:format)	profiles#edit {:format=>:html}
                    PATCH	/profile(.:format)	profiles#update {:format=>:html}
                    PUT	/profile(.:format)	profiles#update {:format=>:html}
                    DELETE	/profile(.:format)	profiles#destroy {:format=>:html}
```
So the path name is `edit_profile_path` with method: delete, right? Except it is not. `DELETE	/profile(.:format)` is unnamed and `edit_profile_path` == `/profile/edit`. Thats why I had to use it as `member`.

Comment (dwradcliffe) [inline config/routes.rb]: You still need member for the get but you don't need `delete :destroy, as: :destroy`

Comment (dwradcliffe) [inline config/routes.rb]: Ohh, nevermind, you're right. That is necessary because the profile_path named route belongs to the other resource.

Comment (qrush) [inline config/routes.rb]: Something to consider: This could be an entirely separate resource and concept in the app. "UserDeletion" perhaps, then it could be handled via a `new` and `create` action and not glommed onto the profile pages.

**Comment by sonalkr132** [comment]
> We don't need a link to contact support about taking over a fully yanked gem

The link is removed. The page is plane old yanked gem page. I really couldn't find any section in guides which explains what can be done with the namespace. Closest one is this page: http://help.rubygems.org/kb/gemcutter/removing-a-published-rubygem

> Let's have one final confirm via JS before going through with the final button.

I have [added data-confirm](https://github.com/rubygems/rubygems.org/pull/1301/files#diff-1c3a6bb16bd767d2d1a7cb109ca2cf43R36). Let me know if this should be something else.

> the existing gems on an account should go through the Deletion process that yanked gems go through.

All the versions of gems where user is only owner will be yanked. Yanking a version is computationally expensive (version is marked unindexed, rubygem's updated_at is changed, version position is changed etc), and we will be yanking multiple of them before user delete. Should we think more about this?

File: app/models/user.rb [hunk 9]
```diff
@@ -5,14 +5,15 @@ class User < ActiveRecord::Base
 
   PERMITTED_ATTRS = [:bio, :email, :handle, :hide_email, :location, :password, :website].freeze
 
+  before_destroy :yank_gems
```
Comment (dwradcliffe) [inline app/models/user.rb]: Wait, why are we yanking gems? I would assume that removing my user does not remove any gems.

File: test/unit/user_test.rb [hunk 10]
```diff
@@ -211,4 +216,24 @@ class UserTest < ActiveSupport::TestCase
...
+      @version = create(:version, rubygem: @rubygem)
+    end
+
+    should "record deletion" do
+      assert_difference 'Deletion.count', 1 do
+        @user.destroy
+      end
+    end
+
+    should "unown rubygem" do
```
Comment (dwradcliffe) [inline test/unit/user_test.rb]: small typo: should be "unknown"

**Comment by dwradcliffe** [comment]
Can we collapse the gem list a little? There's a little too much whitespace in my opinion.

File: app/models/user.rb (hunk 9 above)
Comment (sonalkr132) [inline app/models/user.rb]: Previously, I was just using destroying the records from the database.  @qrush suggested that:
> Instead of destroying gems, the existing gems on an account should go through the Deletion process that yanked gems go through.

If the user being destroyed is the only owner of a gem then that gem will go through the Deletion process. ie it will be yanked.
I suppose it would also automate the process of namespace being taken over?

**Comment by sonalkr132** [comment]
Before:
screenshot from 2016-07-02 19-58-02
After:
screenshot from 2016-07-02 19-58-36

File: app/models/user.rb (hunk 9 above)
Comment (dwradcliffe) [inline app/models/user.rb]: I'm not sure we want that. @qrush can you expand on what you had in mind?

Comment (qrush) [inline app/models/user.rb]: I just figured it would be safer doing a hard deletion of the gems, something we have learned is not a good idea over the years.

We could keep the gems around - but then they are orphaned if the user being deleted is the only author. At least, that's what a quick pass at the `Pusher#authorize` code looks like:

https://github.com/rubygems/rubygems.org/blob/master/app/models/pusher.rb#L23-L24
https://github.com/rubygems/rubygems.org/blob/master/app/models/rubygem.rb#L180

Pusher can authorize a user to push a given gem if it's an unsaved record, has no indexed versions, or it's owned by the user. We don't have a case yet for no ownerships on the gem.

File: app/models/user.rb [hunk 11]
```diff
@@ -14,14 +14,15 @@ class User < ActiveRecord::Base
     :twitter_username
   ].freeze
 
+  before_destroy :yank_gems
```
Comment (sonalkr132) [inline app/models/user.rb]: [Previous discussion](https://github.com/rubygems/rubygems.org/pull/1301#discussion_r69366108) was marked outdated diff on rebase.

It is to be decided whether single owner gems should be yanked on profile delete or a new [rule](https://github.com/rubygems/rubygems.org/blob/master/app/models/pusher.rb#L23-L24) should be added for pushing to orphaned gems.

If we don't yank gem on profile delete and add rule for pushing to orphaned gems, what happens to older versions of orphaned gem? Does new owner becomes owner of previous versions or we yank older versions when the gem is adopted or something else.

Comment (qrush) [inline app/models/user.rb]: I don't think this should change our policy regarding how gems are dealt with. There should always be an owner for a gem. Here's what I think deleting your account should do:

1. For gems with > 1 owner: removes your ownership.
2. For gems with 1 owner (just yourself): yanks all versions of that gem.

We can be very explicit and explain this on a special page as you're deleting your account. It's not an action to be done mistakenly (but that will happen!)

Any objections to this? @dwradcliffe @evanphx @indirect

Comment (dwradcliffe) [inline app/models/user.rb]: I'm okay with that that. Let's make it more clear on the page though. And
maybe include instructions for adding an owner right there too.

On Tuesday, July 12, 2016, Nick Quaranto <notifications@github.com> wrote:

> In app/models/user.rb
> <https://github.com/rubygems/rubygems.org/pull/1301#discussion_r70547314>:
>
> > @@ -14,14 +14,15 @@ class User < ActiveRecord::Base
> >      :twitter_username
> >    ].freeze
> >
> > +  before_destroy :yank_gems
>
> I don't think this should change our policy regarding how gems are dealt
> with. There should always be an owner for a gem. Here's what I think
> deleting your account should do:
>
>    1. For gems with > 1 owner: removes your ownership.
>    2. For gems with 1 owner (just yourself): yanks all versions of that
>    gem.
>
> We can be very explicit and explain this on a special page as you're
> deleting your account. It's not an action to be done mistakenly (but that
> will happen!)
>
> Any objections to this? @dwradcliffe <https://github.com/dwradcliffe>
> @evanphx <https://github.com/evanphx> @indirect
> <https://github.com/indirect>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rubygems/rubygems.org/pull/1301/files/9e00a444a1b185f3d04a63b34113d3171fb786ae#r70547314>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHQQiqQh1C_6ZuJu58v4DAMTyiWjAZdks5qVDJ8gaJpZM4IqAZ2>
> .
>


-- 
---
David Radcliffe

**Comment by sonalkr132** [comment]
I have updated delete page to:
screenshot from 2016-07-14 21-36-28

Let me know the wording you would like to see if this isn't suitable.

File: app/assets/stylesheets/modules/gem.css [hunk 12]
```diff
@@ -31,6 +31,11 @@
       overflow: auto;
       border-bottom: 1px solid #c1c4ca; } }
 
+.delete__owners{
```
Comment (simi) [inline app/assets/stylesheets/modules/gem.css]: Is this valid BEM (delete  = block, owners = element)?

File: app/controllers/api/v1/owners_controller.rb [hunk 13]
```diff
@@ -25,7 +25,8 @@ def create
   def destroy
     owner = @rubygem.owners.find_by_name(params[:email])
     if owner
-      if @rubygem.ownerships.find_by_user_id(owner.id).try(:destroy)
+      ownership = @rubygem.ownerships.find_by(user_id: owner.id)
+      if ownership.safe_destroy
```
Comment (simi) [inline app/controllers/api/v1/owners_controller.rb]: :exclamation: `ownership` can be `nil` here

File: app/controllers/profiles_controller.rb [hunk 14]
```diff
@@ -28,6 +28,21 @@ def update
...
 
+  def delete
+    @only_owner_gems = current_user.only_owner_gems
+    @multi_owner_gems = current_user.rubygems_downloaded - @only_owner_gems
+  end
+
+  def destroy
+    if User.authenticate(current_user.email, params[:user][:password]) && current_user.destroy
+      flash[:notice] = t '.successful_flash'
+      redirect_to_root
```
Comment (simi) [inline app/controllers/profiles_controller.rb]: This can be shortened to one line if you'll avoid `redirect_to_root`, which is just shortcut for `redirect_to root_path`.
```ruby
redirect_to root_path, notice: t '.successful_flash'
```

File: app/models/ownership.rb [hunk 15]
```diff
@@ -4,16 +4,7 @@ class Ownership < ActiveRecord::Base
...
-
-  def keep_last_owner
-    if rubygem.owners.count == 1
-      errors[:base] << "Can't delete last owner of a gem."
-      false
-    else
-      true
-    end
+  def safe_destroy
+    rubygem.owners.count != 1 && destroy ? true : false
```
Comment (simi) [inline app/models/ownership.rb]: What about to implement this as `before_destroy` callback and keep calling `destroy`?

File: app/assets/stylesheets/modules/gem.css (hunk 12 above)
Comment (sonalkr132) [inline app/assets/stylesheets/modules/gem.css]: Thanks. I didn't know about BEM. I have changed `delete_owners` to `owners` block. My reasoning is mostly visual:
screenshot from 2016-07-14 21-58-58

File: app/models/ownership.rb (hunk 15 above)
Comment (sonalkr132) [inline app/models/ownership.rb]: `ownership` deletion is of two types. When profile gets deleted, [last owner of the gem gets deleted](https://github.com/rubygems/rubygems.org/pull/1301/files#diff-4676c008b11a5480d73d4a6de01e45b9R23) as well. However, when a user is removing an owner, we need to check that they are not removing the only owner. Callback would run on all the destroys, which is not what we need.

I know this is ugly but sadly this is the solution I have. Let me know if you can think of something better.

Comment (simi) [inline app/models/ownership.rb]: Ok, that makes sense. What about to rename this to something better describing the usage?

You can move

```ruby
rubygem.owners.count != 1
```
somewhere else to just call something like

```ruby
rubygem.many_owners? && destroy
```

Also is there chance that gem has 0 owners?

I'm not sure why `? true : false` part is needed since it already returns `true`/`false` AFAIK.

Comment (sonalkr132) [inline app/models/ownership.rb]: > Also is there chance that gem has 0 owners?

No.
I have added `many_owners?` function and removed redundant ternary operator. Thanks.

File: app/models/rubygem.rb [hunk 16]
```diff
@@ -119,6 +119,10 @@ def unowned?
     ownerships.blank?
   end
 
+  def many_owners?
+    owners.count > 1
```
Comment (simi) [inline app/models/rubygem.rb]: Ahh, sorry I forgot to mention. There already [many?](https://github.com/rails/rails/blob/0800e5dcad571c00c8c33e886c50529001d0badc/activerecord/lib/active_record/associations/collection_association.rb#L367) method on collection association relations. It will be enough to delegate it or use here.

Comment (sonalkr132) [inline app/models/rubygem.rb]: Great :+1:

File: config/routes.rb (hunk 6 above)
Comment (simi) [inline config/routes.rb]: @qrush I like that idea, same approach is (was) used for yank/unyank => `deletions#create/destroy` [routes](https://github.com/rubygems/rubygems.org/blob/d6a762018a10af25f7180ee1018bb643c318fd3a/config/routes.rb#L74).

Comment (sonalkr132) [inline config/routes.rb]: I may be failing to see the appeal of a separate controller. Can someone please explain? `ProfilesController` handles `edit` and `update` of profiles, I would believe that `delete` and `destroy` are related concepts. The controller isn't large either.
Currently routes look like:
```Ruby
delete_profile    GET     /profile/delete(.:format)    profiles#delete {:format=>:html}
destroy_profile   DELETE  /profile(.:format)           profiles#destroy {:format=>:html}
edit_profile      GET     /profile/edit(.:format)      profiles#edit {:format=>:html}
                  PATCH   /profile(.:format)           profiles#update
```
I am sorry but how is this non intuitive?

**Comment by indirect** [comment]
I am happy with this. It seems like all the outstanding comments have been addressed (and it seems okay to me that deleting an account and yanking all owned gems could take some time). @sonalkr132, do you think this could take more than 30 seconds? that would cause this request to be timed out by puma or nginx.

**Comment by sonalkr132** [comment]
>  It seems like all the outstanding comments have been addressed

It was suggested that deletion of users could be a separate resource(the way deletion is a separate resource for gems). Someone is yet to explain need for it :grin: ([Discussion link](https://github.com/rubygems/rubygems.org/pull/1301#discussion_r66561590))

>  do you think this could take more than 30 seconds?

Yeah. 30 seconds is barely enough to yank 30 versions. May be we can move yanking of gems to background process?

**Comment by indirect** [comment]
I think it is OK to delete users via the existing resources. If we come up with useful reasons to make it a separate resource later, we can refactor then.

I am okay with removing gems in the background, but we should probably wait to delete the account until all of the gems have actually been removed—it seems bad to delete the account before the background jobs to remove the gems have run. Maybe we can have an additional background job that tries to delete the account itself, checking every few seconds until all the owned gems have been yanked?

**Comment by dwradcliffe** [comment]
Let's just move the whole process to one background job. Then we can say that it has been scheduled for deletion and in reality will probably only be a minute or two.

**Comment by indirect** [comment]
sounds great 👍

**Comment by indirect** [comment]
@sonalkr132 normally, background jobs do not serialize the object at all. Instead, pass the object ID, and use ActiveRecord to find a new copy of the object inside the background job before doing operations on it. For example:

```ruby
class DeleteUserJob
  def perform(user_id)
    user = User.find(user_id)
  rescue ActiveRecord::RecordNotFound
    # already deleted, hooray
  end
end
```

**Comment by sonalkr132** [comment]
I have updated it to destroy user in background job. User will get a confirmation mail of successful deletion once the job gets processed.

File: app/assets/stylesheets/modules/gem.css [hunk 17]
```diff
@@ -31,6 +31,11 @@
       overflow: auto;
       border-bottom: 1px solid #c1c4ca; } }
 
+.owners{
```
Comment (segiddins) [inline app/assets/stylesheets/modules/gem.css]: doesn't seem to use the same formatting as the surrounding declarations

Comment (sonalkr132) [inline app/assets/stylesheets/modules/gem.css]: I have fixed the positioning of curly brackets. If that is not what you meant, please elaborate :)

File: app/controllers/api/v1/owners_controller.rb [hunk 18]
```diff
@@ -25,8 +25,9 @@ def create
   def destroy
     owner = @rubygem.owners.find_by_name(params[:email])
     if owner
-      if @rubygem.ownerships.find_by_user_id(owner.id).try(:destroy)
-        render plain: "Owner removed successfully."
+      ownership = @rubygem.ownerships.find_by(user_id: owner.id)
+      if ownership&.safe_destroy
```
Comment (evanphx) [inline app/controllers/api/v1/owners_controller.rb]: Please don't use this syntax because it only exists in very latest ruby's which I'd rather we not depend implicitly on.

File: app/models/user.rb (hunk 11 above)
Comment (evanphx) [inline app/models/user.rb]: I agreed, deleting a user should not yank all their gems if there are multiple owners.

File: app/assets/stylesheets/modules/gem.css [hunk 19]
```diff
@@ -31,6 +31,10 @@
       overflow: auto;
       border-bottom: 1px solid #c1c4ca; } }
 
+.owners {
```
Comment (simi) [inline app/assets/stylesheets/modules/gem.css]: Is this valid BEM?

File: app/assets/stylesheets/modules/gem.css (hunk 17 above)
Comment (segiddins) [inline app/assets/stylesheets/modules/gem.css]: 👌🏻

File: app/controllers/profiles_controller.rb [hunk 20]
```diff
@@ -29,6 +29,17 @@ def update
     end
   end
 
+  def delete
+    @only_owner_gems = current_user.only_owner_gems
+    @multi_owner_gems = current_user.rubygems_downloaded - @only_owner_gems
+  end
+
+  def destroy
+    Delayed::Job.enqueue DeleteUser.new(current_user), priority: PRIORITIES[:web_hook]
```
Comment (simi) [inline app/controllers/profiles_controller.rb]: `priority: PRIORITIES[:web_hook]`

What about to create new priority for this action (even with same value as :web_hook)?

File: app/jobs/delete_user.rb [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def perform
+    profile = User.find(user['id'])
+    email = profile.email
+    profile.destroy
```
Comment (simi) [inline app/jobs/delete_user.rb]: Is it possible this returns false and account will not be destroyed, but the email will be sent anyway?

File: app/mailers/mailer.rb [hunk 22]
```diff
@@ -17,4 +17,10 @@ def email_confirmation(user)
          subject: I18n.t('mailer.confirmation_subject',
            default: 'Please confirm your email address with RubyGems.org')
   end
+
+  def deletion_confirmation(email)
```
Comment (simi) [inline app/mailers/mailer.rb]: `confirmation` sounds like you need to confirm something to me. If I understand this well, this is just info email.

File: app/views/profiles/_rubygem.html.erb [hunk 23]
```diff
@@ -1,10 +1,10 @@
 <li class="<%= "small" unless graph %>">
-  <a href="<%= rubygem_path(rubygem) %>" class="gems__gem">
+  <div class="gems__gem">
     <span class="gems__gem__info">
-      <p class="gems__gem__name">
+      <a href="<%= rubygem_path(rubygem) %>" class="gems__gem__name">
```
Comment (simi) [inline app/views/profiles/_rubygem.html.erb]: Wondering why `link_to` wasn't used here before.

File: app/controllers/api/v1/owners_controller.rb (hunk 18 above)
Comment (sonalkr132) [inline app/controllers/api/v1/owners_controller.rb]: Although I have switched back to using `try` here, we have been using safe navigation operator for a quite a while now. I doubt codebase supports older rubies.

File: app/views/profiles/edit.html.erb [hunk 24]
```diff
@@ -84,3 +84,9 @@
 <% end %>
 
 <%= button_to t('.api_access.reset'), reset_api_v1_api_key_path, :method => :put, :data => {:confirm => t('.api_access.confirm_reset')}, :class => 'form__submit' %>
+
+<div class="t-body">
+  <h2 class="delete_header"><%= t '.delete.delete_profile' %></h2>
```
Comment (simi) [inline app/views/profiles/edit.html.erb]: Is `delete_header` class needed here? I wasn't able to find it in CSS.

File: app/assets/stylesheets/modules/gem.css (hunk 19 above)
Comment (sonalkr132) [inline app/assets/stylesheets/modules/gem.css]: I would argue that yes it is :) `.owners` is a block and `.gem__owners` are its elements. Possibly it could be better named. I am open to suggestions :email:

Comment (simi) [inline app/assets/stylesheets/modules/gem.css]: I'm not BEM expert. I was just curious. Feel free to leave it as is.

File: app/jobs/delete_user.rb (hunk 21 above)
Comment (sonalkr132) [inline app/jobs/delete_user.rb]: When you say it like that it does seem it can use a conditional :sweat_smile: 
What would you suggest? Should we just not deliver the confirmation mail when it fails or should we send one informing that it failed?

File: app/views/profiles/_rubygem.html.erb (hunk 23 above)
Comment (sonalkr132) [inline app/views/profiles/_rubygem.html.erb]: It intends to mark complete following section as link:
```Ruby
<%= rubygem.name %>
<span class="gems__gem__version"><%= rubygem.versions.most_recent %></span>
```
which wouldn't be as easy with `link_to`.

**Comment by sonalkr132** [comment]
@simi I have incorporated most of your suggestions :+1: Please let me know if you are satisfied with changes made.

File: app/views/profiles/_rubygem.html.erb (hunk 23 above)
Comment (simi) [inline app/views/profiles/_rubygem.html.erb]: But it is not problem now, is it?

Comment (sonalkr132) [inline app/views/profiles/_rubygem.html.erb]: Yes it is. Following is what it looks like on this PR:
```Ruby
<a href="<%= rubygem_path(rubygem) %>" class="gems__gem__name">
   <%= rubygem.name %>
   <span class="gems__gem__version"><%= rubygem.versions.most_recent %></span>
</a>
```

Comment (simi) [inline app/views/profiles/_rubygem.html.erb]: Damn, I'm blind. Sorry :(
